From 858213a82a0beb388e5cbd5dc5183a80aa18806e Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Thu, 18 Jun 2026 20:22:14 +0200
Subject: [PATCH 6/6] loader/win32u: Install bundled fonts with full registry
 paths.

Some games expect those registry entries to point somewhere real, so copy Wine's bundled fonts into the prefix
and register them as full C:\windows\Fonts paths instead of plain filenames like msyh.ttf or tahoma.ttf.

Also fixes update_external_font_keys() deleting the registry entries on every subsequent process start.
---
 dlls/win32u/font.c | 19 +++++++++++++++++--
 loader/wine.inf.in | 38 ++++++++++++++++++++++++++------------
 2 files changed, 43 insertions(+), 14 deletions(-)

diff --git a/dlls/win32u/font.c b/dlls/win32u/font.c
index 2134f71db95..ece483e03a7 100644
--- a/dlls/win32u/font.c
+++ b/dlls/win32u/font.c
@@ -6672,8 +6672,23 @@ static void update_external_font_keys(void)
 
     LIST_FOR_EACH_ENTRY_SAFE( key, next, &external_keys, struct external_key, entry )
     {
-        reg_delete_value( win9x_key, key->value );
-        reg_delete_value( winnt_key, key->value );
+        WCHAR *face_name = wcsdup( key->value );
+        struct gdi_font_face *int_face = NULL;
+
+        if (face_name)
+        {
+            WCHAR *suffix;
+            if ((suffix = wcsrchr( face_name, ' ' )) && !facename_compare( suffix, true_type_suffixW, -1 ))
+                *suffix = 0;
+            int_face = find_face_from_full_name( face_name );
+            free( face_name );
+        }
+        /* face moved to internal (windows fonts dir); keep registry entry, only drop cache */
+        if (!int_face || (int_face->flags & ADDFONT_EXTERNAL_FONT))
+        {
+            reg_delete_value( win9x_key, key->value );
+            reg_delete_value( winnt_key, key->value );
+        }
         reg_delete_value( hkey, key->value );
         list_remove( &key->entry );
         free( key );
diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index f010afe2c91..e697359ad94 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -54,7 +54,7 @@ AddService=MountMgr,0x800,MountMgrService
 RegisterDlls=RegisterDllsSection
 WineFakeDlls=FakeDlls
 UpdateInis=SystemIni
-CopyFiles=ColorFiles,EtcFiles,InfFiles,NlsFiles,SortFiles,WinmdFiles
+CopyFiles=ColorFiles,EtcFiles,InfFiles,NlsFiles,SortFiles,WinmdFiles,FontFiles
 AddReg=\
     Classes,\
     ContentIndex,\
@@ -120,7 +120,7 @@ AddReg=\
 [BaseWow64Install]
 RegisterDlls=RegisterDllsSection
 WineFakeDlls=FakeDllsWin32,FakeDllsWow64
-CopyFiles=NlsFiles,WinmdFiles
+CopyFiles=NlsFiles,WinmdFiles,FontFiles
 AddReg=\
     CurrentVersion,\
     Debugger,\
@@ -537,26 +537,26 @@ HKLM,%FontSubStr%,"Times New Roman CYR,204",,"Times New Roman,204"
 HKLM,%FontSubStr%,"Times New Roman Greek,161",,"Times New Roman,161"
 HKLM,%FontSubStr%,"Times New Roman TUR,162",,"Times New Roman,162"
 HKLM,System\CurrentControlSet\Hardware Profiles\Current\Software\Fonts,"LogPixels",0x10003,0x00000060
-HKLM,%FontsNT%,"Arial (TrueType)",,"arial.ttf"
+HKLM,%FontsNT%,"Arial (TrueType)",,"C:\windows\Fonts\arial.ttf"
 HKLM,%FontsNT%,"Arial Bold (TrueType)",,"arialbd.ttf"
-HKLM,%FontsNT%,"Courier New (TrueType)",,"cour.ttf"
+HKLM,%FontsNT%,"Courier New (TrueType)",,"C:\windows\Fonts\cour.ttf"
 HKLM,%FontsNT%,"Courier New Bold (TrueType)",,"courbd.ttf"
 HKLM,%FontsNT%,"Malgun Gothic (TrueType)",,"malgun.ttf"
-HKLM,%FontsNT%,"Marlett (TrueType)",,"marlett.ttf"
+HKLM,%FontsNT%,"Marlett (TrueType)",,"C:\windows\Fonts\marlett.ttf"
 HKLM,%FontsNT%,"Microsoft Sans Serif (TrueType)",,"micross.ttf"
 HKLM,%FontsNT%,"MS Gothic (TrueType)",,"msgothic.ttc"
 HKLM,%FontsNT%,"MS PGothic (TrueType)",,"msgothic.ttc"
 HKLM,%FontsNT%,"MS UI Gothic (TrueType)",,"msgothic.ttc"
-HKLM,%FontsNT%,"Microsoft YaHei (TrueType)",,"msyh.ttf"
+HKLM,%FontsNT%,"Microsoft YaHei (TrueType)",,"C:\windows\Fonts\msyh.ttf"
 HKLM,%FontsNT%,"Nirmala UI (TrueType)",,"nirmala.ttf"
 HKLM,%FontsNT%,"SimSun (TrueType)",,"simsun.ttc"
 HKLM,%FontsNT%,"NSimSun (TrueType)",,"simsun.ttc"
-HKLM,%FontsNT%,"Symbol (TrueType)",,"symbol.ttf"
-HKLM,%FontsNT%,"Tahoma (TrueType)",,"tahoma.ttf"
-HKLM,%FontsNT%,"Tahoma Bold (TrueType)",,"tahomabd.ttf"
-HKLM,%FontsNT%,"Times New Roman (TrueType)",,"times.ttf"
-HKLM,%FontsNT%,"Webdings (TrueType)",,"webdings.ttf"
-HKLM,%FontsNT%,"Wingdings (TrueType)",,"wingdings.ttf"
+HKLM,%FontsNT%,"Symbol (TrueType)",,"C:\windows\Fonts\symbol.ttf"
+HKLM,%FontsNT%,"Tahoma (TrueType)",,"C:\windows\Fonts\tahoma.ttf"
+HKLM,%FontsNT%,"Tahoma Bold (TrueType)",,"C:\windows\Fonts\tahomabd.ttf"
+HKLM,%FontsNT%,"Times New Roman (TrueType)",,"C:\windows\Fonts\times.ttf"
+HKLM,%FontsNT%,"Webdings (TrueType)",,"C:\windows\Fonts\webdings.ttf"
+HKLM,%FontsNT%,"Wingdings (TrueType)",,"C:\windows\Fonts\wingding.ttf"
 HKCU,Software\Wine\Fonts\Replacements,"Palatino Linotype",,"Times New Roman"
 HKCU,Software\Wine\Fonts\Replacements,"Verdana",,"Times New Roman"
 HKCU,Software\Wine\Fonts\Replacements,"Segoe UI Semibold",,"Times New Roman"
@@ -1571,6 +1571,18 @@ normnfkd.nls
 [SortFiles]
 sortdefault.nls
 
+[FontFiles]
+arial.ttf
+cour.ttf
+marlett.ttf
+msyh.ttf
+symbol.ttf
+tahoma.ttf
+tahomabd.ttf
+times.ttf
+webdings.ttf
+wingding.ttf
+
 [WinmdFiles]
 windows.applicationmodel.winmd
 windows.globalization.winmd
@@ -1586,6 +1598,7 @@ windows.ui.xaml.winmd
 [WineSourceDirs]
 NlsFiles  = nls
 SortFiles = nls
+FontFiles = fonts
 WinmdFiles = winmd,include
 
 [DestinationDirs]
@@ -1594,6 +1607,7 @@ EtcFiles  = 12,etc
 InfFiles  = 17
 NlsFiles  = 11
 SortFiles = 10,globalization\sorting
+FontFiles = 20
 WinmdFiles = 11,winmetadata
 
 [SteamClient]
-- 
2.54.0

